File: 0239-sliding-window-maximum/0239-sliding-window-maximum.java
class Solution {
    class MyDeque{
        Deque<Integer> queue = new LinkedList<>();

        void poll(int num) {
            if(!queue.isEmpty() && queue.peek() == num) {
                queue.poll();
            }
        }

        void add(int num) {
            while(!queue.isEmpty() && num > queue.getLast()){
                queue.removeLast();
            }
            queue.add(num);
        }

        int peek(){
            return queue.peek();
        }
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        if(nums.length == 1) {
            return nums;
        }

        MyDeque queue = new MyDeque();

        int[] arr = new int[nums.length - k + 1];
        int count = 0;
        for (int i = 0; i < k; i++) {
            queue.add(nums[i]);
        }
        arr[count++] = queue.peek();

        for(int i = k; i < nums.length; i++) {
            queue.poll(nums[i - k]);
            queue.add(nums[i]);
            arr[count++] = queue.peek();
        }
        return arr;
    }
}

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque
class MyQueue:
    def __init__(self):
        self.queue = deque()

    def push(self, val):
        if not self.queue:
            self.queue.append(val)
        else:
            while self.queue and val > self.queue[-1]:
                self.queue.pop()

            self.queue.append(val)
    
    def remove(self, val):
        if self.queue[0] == val:
            self.queue.popleft()

    def front(self):
        return self.queue[0]


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = MyQueue()
        res = []

        for i in range(len(nums)):
            queue.push(nums[i])

            if i >= k:
                queue.remove(nums[i - k])

            if i >= k - 1:
                res.append(queue.front())

        return res
            
            

